service/history/queueProcessorFactory.go

```diff
@@ -0,0 +1,156 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package history
+
+import (
+	sdkclient "go.temporal.io/sdk/client"
+	"go.temporal.io/server/api/historyservice/v1"
+	"go.temporal.io/server/common/persistence/visibility/manager"
+	"go.temporal.io/server/common/resource"
+	"go.temporal.io/server/service/history/queues"
+	"go.temporal.io/server/service/history/shard"
+	"go.temporal.io/server/service/history/workflow"
+	"go.temporal.io/server/service/worker/archiver"
+	"go.uber.org/fx"
+)
+
+var QueueProcessorModule = fx.Options(
+	fx.Provide(
+		fx.Annotated{
+			Group:  queues.ProcessorFactoryFxGroup,
+			Target: NewTransferQueueProcessorFactory,
+		},
+		fx.Annotated{
+			Group:  queues.ProcessorFactoryFxGroup,
+			Target: NewTimerQueueProcessorFactory,
+		},
+		fx.Annotated{
+			Group:  queues.ProcessorFactoryFxGroup,
+			Target: NewVisibilityQueueProcessorFactory,
+		},
```

Comment:
alternatively, you can use fx.Out like:
```
type queueFactoryResult struct {
  fx.Out
  QueueFactories []queues.ProcessorFactory `group:"queueProcessorFactory,flatten"`
}
```
Then you just need one provider that provides the queueFactoryResult where you create all factories.

Reply:
With one provider, we will need to pass the dependencies for all queue processors into that provider, which makes me feel that the dependency for each queue processor if not very clear. But I don't have a strong opinion on that.

Reply:
You can have that one provider take multiple inputs (each for one queue type with fx.In embedded). But it is up to you.